### PR TITLE
TASK: Process all resources in ThumbnailGenerator

### LINKED ISSUE #1469: Allow forcing of thumbnail generation in ThumbnailService
### Description

Currently the `ThumbnailService` returns the original image if the size is not changed or a larger image is requested than the original.

This makes it impossible to optimize images via the `Moc/ImageOptimizer` package (https://github.com/mocdk/MOC.ImageOptimizer/issues/14) and similar techniques, as the method `refreshThumbnail` is never called and cannot be hooked into to do post-processing.

The solution from #1178 is not enough to solve this as it also doesn't allow using custom libraries and behavior.

#### Expected behavior

The `ThumbnailService` should have a option to always force the generation of thumbnails and therefore allow post-processing of all images.

#### Actual behavior

1. Create thumbnail from image which is same or larger than the one requested
2. The original unmodified image is returned and no thumbnail variant generated

### Affected Versions

Neos: 

2.3
3.0

### PR DESCRIPTION
**Pull request by Sebobo**
This change allows to modify the processing behavior in custom ``ThumbnailGenerator`` implementations.

The ``ThumbnailService`` will not check the size of the image anymore and return the asset in some cases without triggering the processing pipeline.

This only increases computation a bit as the ``ResizeImageAdjustment`` will also check for the image dimensions and do nothing if they don’t change. Which is the same as the ``ThumbnailService`` previously did.

Previously images which were smaller than the requested image size were not processed and didn't trigger the thumbnail refresh event.

Resolves #1469

### CODE REVIEW

File: Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php [hunk 1]
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Comment (aertmann) [inline Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php]: why this change?

File: Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php [hunk 2]
```diff
@@ -80,4 +81,14 @@ public function refresh(Thumbnail $thumbnail)
...
+    /**
+     * @param PersistentResource $resource
+     * @param array $adjustments
+     * @return array
+     */
+    protected function processResource(PersistentResource $resource, array $adjustments)
+    {
+        return $this->imageService->processImage($resource, $adjustments);
+    }
```
Comment (aertmann) [inline Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php]: what does this method help with?

File: Neos.Media/Classes/Domain/Service/ThumbnailService.php [hunk 3]
```diff
@@ -92,11 +92,6 @@ public function getThumbnail(AssetInterface $asset, ThumbnailConfiguration $conf
             if ($asset->getWidth() === null && $asset->getHeight() === null) {
                 return $asset;
             }
-            $maximumWidth = ($configuration->getMaximumWidth() > $asset->getWidth()) ? $asset->getWidth() : $configuration->getMaximumWidth();
-            $maximumHeight = ($configuration->getMaximumHeight() > $asset->getHeight()) ? $asset->getHeight() : $configuration->getMaximumHeight();
-            if ($configuration->isUpScalingAllowed() === false && $maximumWidth === $asset->getWidth() && $maximumHeight === $asset->getHeight()) {
-                return $asset;
-            }
```
Comment (aertmann) [inline Neos.Media/Classes/Domain/Service/ThumbnailService.php]: so just remove that feature completely instead of making it configurable/overridable?

File: Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php (hunk 2 above)
Comment (Sebobo) [inline Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php]: @aertmann It can be easily overriden with custom generators.
This way we could for example use a custom generator instead of the AOP in the ImageOptimizer package.

File: Neos.Media/Classes/Domain/Service/ThumbnailService.php (hunk 3 above)
Comment (Sebobo) [inline Neos.Media/Classes/Domain/Service/ThumbnailService.php]: @aertmann I first made it configurable but then saw that a check happens anyway at a later time. The code gets quite bulky when we add additional option and I see it actually as a bug if Neos decides here that images don't need any processing at all only based on the image size.

File: Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php (hunk 1 above)
Comment (Sebobo) [inline Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php]: Because we have empty lines there in all other files and it's default formatting

File: Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php (hunk 2 above)
Comment (Sebobo) [inline Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php]: And this way it's also easier to copy the resulting image and create an alternative if we don't want to modify the original.

File: Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php (hunk 1 above)
Comment (aertmann) [inline Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php]: uhm no we don't, just checked 10 random ones and they all didn't.. maybe you're using some kind of special rules normally?

File: Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php (hunk 2 above)
Comment (aertmann) [inline Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php]: Well if you wanna change generator, you're supposed to create a different generator, not AOP this one. That's the whole point of having them.

File: Neos.Media/Classes/Domain/Service/ThumbnailService.php (hunk 3 above)
Comment (aertmann) [inline Neos.Media/Classes/Domain/Service/ThumbnailService.php]: there's another place that does the same? where?

Well that's definitely a matter of perspective, I'm sure when this was first implemented it was meant as a feature to prevent generating identical thumbnails. I personally would prefer it wasn't here, but would some more opinions on this.

don't see why the code would become bloated, aside that this method already is quite bloated itself.. but could be moved to a separate method and the simply have a configuration injected boolean property checking..

File: Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php (hunk 2 above)
Comment (Sebobo) [inline Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php]: That's what I'm saying. We don't need the AOP of the refresh method anymore like previously. 
But without this change in a custom generator you need to override a lot more code. Now you only need to override `processResource` in most cases and you can do what you want.

Comment (aertmann) [inline Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php]: well I'd still do it that way instead of providing a different thumbnail converter since it's basically a add-on needed after the fact.. anyway that's irrelevant for this change though.

okay so to extend the thumbnail generator but change the generator, I see, still the method by itself it a little odd

File: Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php (hunk 1 above)
Comment (kdambekalns) [inline Neos.Media/Classes/Domain/Model/ThumbnailGenerator/ImageThumbnailGenerator.php]: `namespace` is the first line that's our rule. No leading empty line.

File: Neos.Media/Classes/Domain/Service/ThumbnailService.php (hunk 3 above)
Comment (Sebobo) [inline Neos.Media/Classes/Domain/Service/ThumbnailService.php]: The `ImageResizeAdjustment` also checks in a few places if the size can/should be changed or not. 
The result of my change would of course create separate db entries for new thumbnails but wouldn't create new images if the sha1 doesn't change.

I'm not sure if there is a use case where this could be a actual problem. If so, then this should be adjusted further.

Having a global option `forceThumbnailGeneration` isn't quite right, as it doesn't actually force the creation of a separate image without further code changes. 

I'm working on a actual new feature which allows to extend the `adjustments` array with custom post-processors. This then allows image optimization, filtering and whatever the integrator wants on a per image basis. But for that all images have to go into the refresh function and every adjustment has to decide by itself if it's applied or not.

Of course a `forceImageProcessing` global config could help there too, but it didn't feel right for me.
So more opinions would be great.

**Comment by Sebobo** [comment]
The whole functionality about the thumbnail generation definitely needs a rewrite in some parts. But that's something for a later version for which I'm glad to help with.
Currently we have the problem that not all images are processed reliably and a editor or integrator cannot rely on an automatic optimization like the ImageOptimizer package does it now.
I'm fine with retargeting the PR to a 3.x version. 

Having a condition like here https://github.com/neos/neos-development-collection/pull/1178/files#diff-16e8cf609f0f20b6ffe12ce13b5f5679L97 doesn't feel right for me. I tried the same at first.

File: Neos.Media/Classes/Domain/Service/ThumbnailService.php (hunk 3 above)
Comment (gerhard-boden) [inline Neos.Media/Classes/Domain/Service/ThumbnailService.php]: In case we decide to remove those lines, the doc block needs to be updated https://github.com/neos/neos-development-collection/pull/1480/files#diff-16e8cf609f0f20b6ffe12ce13b5f5679L79

**Comment by gerhard-boden** [comment]
@Sebobo thanks for looking into this! As you already said this must go into master, because of the changes in behaviour. 

I really like to have a solution for the described problem, but since this change is not optional what about this scenario: I have 1.000.000 images which I currently deliver in original size (so no thumbnail generation needed). Wouldn't I be left with 2.000.000 images after applying this change? What would my options so I don't have to generate and store 1.000.000 additional images?

**Comment by Sebobo** [comment]
@gerhard-boden Neos doesn't create separate images when no change is actually happening but you would have 1.000.000 additional db entries for the thumbs. Maybe there is another easy way to keep that from happening that I haven't seen yet.

As I said, the in my point of view bigger solution is to open up the image adjustments array to the outside and send each image through the pipeline. If no adjustment is interested in changing the image, we don't need the thumbnail. 
I can also close this PR and work on finishing the bigger one with the mentioned functionality but as it's also a bigger change this will make things more complicated. And at the end it will also need to get rid of this condition. 

Love to get more feedback and ideas!

**Comment by gerhard-boden** [comment]
@Sebobo Thanks for clearing thing up, 1.000.000 DB entries are not half as bad as 1.000.000 duplicate files 😃 But I still don't understand why a optional setting and a condition would be so bad? Depending if you want to fire up the processing pipeline for every image you could tweak the setting. And plugins like the `ImageOptimizer` could use the configuration as well.

**Comment by Sebobo** [comment]
I'll come up with something new and close this for now.
